Create command_sensor.py

Bart274 opened the pull request:
This adds a binary command_sensor

Bart274 committed:
Create command_sensor.py

This adds a binary command_sensor

fabaff commented on homeassistant/components/binary_sensor/command_sensor.py:
```diff
@@ -0,0 +1,101 @@
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.command_sensor/
+"""
```
It would be nice of you update the header.

fabaff commented on homeassistant/components/binary_sensor/command_sensor.py:
```diff
@@ -0,0 +1,101 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = "Command Sensor"
```
Default name is the same as for the `command_sensor` ;-)

fabaff commented:
Please update the .coveragerc file to exclude your platform. `pylint` is complaining:
```
************* Module homeassistant.components.binary_sensor.command_sensor
E: 77,28: Instance of 'CommandBinarySensor' has no 'value_template' member (no-member)
```

Bart274 committed:
Update command_sensor.py

Bart274 committed:
Update .coveragerc

Bart274 commented:
Are these changes better? :)

balloob commented:
CI reports some style violations

On Wed, Jan 13, 2016, 03:21 Bart274 <notifications@github.com> wrote:

> Are these changes better? :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/889#issuecomment-171258904>
> .
>

rmkraus commented on homeassistant/components/binary_sensor/command_sensor.py:
```diff
@@ -0,0 +1,98 @@
+class CommandBinarySensor(BinarySensorDevice):
+    """ Represents a binary sensor that is returning a value of a shell commands. """
+    def __init__(self, hass, data, name, payload_on, payload_off, value_template):
+        self._hass = hass
```
Should be `self.hass` and you probably don't need to set it here. `self.hass` will be set by the `add_devices` callback in `setup_platform`.

Bart274 replied:
are you sure about this? I looked at the mqtt binary sensor and it does the same thing...

rmkraus commented on homeassistant/components/binary_sensor/command_sensor.py:
```diff
@@ -0,0 +1,98 @@
+
+
+# pylint: disable=too-few-public-methods
+class CommandSensorData(object):
```
This class is unnecessary. the update function can be moved into your `CommandBinarySensor` class without sacrificing anything. Classes with only one method are rarely necessary.

Bart274 replied:
I created this component based on the command_sensor in the regular sensor section. It does the same thing, so I kept it this way to be uniform with that.

balloob replied:
Instead of copy pasting it into this file, please import it from the command sensor.

```python
from homeassistant.component.sensor.command_sensor import CommandSensorData
```

happyleavesaoc commented on homeassistant/components/binary_sensor/command_sensor.py:
```diff
@@ -0,0 +1,98 @@
+        _LOGGER.info('Running command: %s', self.command)
+
+        try:
+            return_value = subprocess.check_output(self.command, shell=True)
```
I suggest that this component support state in the same way that the `command_switch` does (both return code and return value). Indeed, the functionality is so similar it might be a good idea to inherit from the same base class (platform `Command`?)

balloob replied:
I agree that we should use inheritance. I already did the same for the `rest` binary_sensor and have a story for `aRest`.

Bart274 replied:
I'll take a look at how you did it with the rest sensors. I haven't based this code on the command_switch, but on the command_sensor.

Bart274 committed:
Update command_sensor.py

Bart274 commented:
@balloob some of the style issues should be solved now

Bart274 committed:
Update command_sensor.py

balloob commented on .coveragerc:
```diff
@@ -41,6 +41,7 @@ omit =
 
     homeassistant/components/binary_sensor/arest.py
     homeassistant/components/binary_sensor/rest.py
+    homeassistant/components/binary_sensor/command_sensor.py
```
This is a pure Python component and should not be excluded from coverage. Please remove this line.

Bart274 replied:
I added because @fabaff told me to do it two days ago...

fabaff replied:
Sorry, my bad. I suggested to do that.

Bart274 committed:
Update .coveragerc

This shouldn't be excluded according to @balloob because it's a pure python component

Bart274 committed:
Update command_sensor.py

Importing CommandSensorData from the command_sensor in sensors in order not to duplicate code

fabaff commented:
Initial docs: https://github.com/balloob/home-assistant.io/commit/3cd3ebed40085be135a1432bee12e898e2d4da8e
Feedback would be welcome. Thanks.

Bart274 commented:
@fabaff & @balloob any idea what caused this build to fail?

balloob commented on homeassistant/components/binary_sensor/command_sensor.py:
```diff
@@ -0,0 +1,82 @@
+
+from homeassistant.const import CONF_VALUE_TEMPLATE
+from homeassistant.components.binary_sensor import BinarySensorDevice
+from homeassistant.component.sensor.command_sensor import CommandSensorData
```
`components` should be plural.

Please make sure you test your code locally before pushing it out. This could have been caught easily.

Bart274 replied:
I can't believe I missed that

Bart274 committed:
Update command_sensor.py

Bart274 committed:
Update command_sensor.py

Bart274 commented:
Finally both checks passed :)

balloob commented:
Woohoo :dolphin:

Bart274 commented:
my first official component :)
